Land V8 4.6 onto master

ofrobots opened the pull request:
Now that V8 4.6 has [hit stable](http://googlechromereleases.blogspot.com/2015/10/stable-channel-update.html), this PR brings-in the latest V8 4.6 patch level (4.6.85.25) after rebasing `vee-eight-4.6` onto `master` as of this morning. **ALL** of the commits on this PR need to be reviewed as they were not reviewed when landing on `vee-eight-4.6`.

Ref: #2688 
R=@bnoordhuis,@nodejs/v8

indutny commented:
LGTM, if CI is green

ofrobots commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/498/

domenic commented:
I am a little concerned about how many commits we are back-porting from future V8 instead of just using the official stable version branches. I don't have any concrete feedback but I think it's something we should try to minimize in general. I am counting 8 of them here?

indutny commented:
@domenic suggestion? Having slow and broken v8?

domenic commented:
Well, as I said I don't have any concrete feedback. But if these patches are not important enough for V8 to backport to 4.6 then the resulting V8 isn't going to be that slow and broken. Unless you are saying Chrome 46 is slow and broken.

It's entirely possible that every single one of these 8 patches is truly necessary, is more necessary for Node.js use cases than for Chrome use cases, and is low-risk. I haven't reviewed them. But it seems like something to be careful about. After all, at some point you might as well just use V8 4.7 instead, if V8 4.6 is so unsuitable for you.

indutny commented:
@domenic oh certainly. I think we will be in a good shape when 4.7 will be stable, until now these patches are necessary for node to run fast.

trevnorris commented:
@domenic 
> 	doc: update V8 options in man page

Update `--v8-options`. Nothing serious.

> 	deps: backport 56a0a79 from V8 upstream

Export post-mortem data for Node. Doubt they'd take the time to back port this.

> 	deps: backport 0d01728 from v8's upstream

Issue that caused our internal use of ArrayBuffers to crash on GC. Don't know if this is important enough for Chrome, but we can't function without it.

> 	deps: improve ArrayBuffer performance in v8

This one is a plain performance issue for v8. They should have back ported it by now.

> 	test: fix tests after V8 upgrade 

Update to our tests. Nothing directly to do with v8.

> 	deps: cherry-pick backports to V8

Deals with `--abort-on-uncaught-exception`, post-mortem data and a performance fix. Most of these they probably don't care about, but don't know why they haven't back ported the performance patch.

bnoordhuis commented:
LGTM but can you update the Reviewed-By tags in the backported patches?  The backports themselves look correct to me.

I echo @domenic's sentiment that the delta is pretty big right now.  Remember that it's possible to CL cherry-picks with the tools/release/merge_to_branch.py script.

I think the postmortem stuff, ArrayBuffer fixes and performance fixes are good candidates.  SetAbortOnUncaughtExceptionCallback() probably isn't eligible because it's an API change.

trevnorris commented:
Also, is it necessary to have the `deps: upgrade V8 to 4.6.85.23` commit?

ofrobots commented:
I agree with the sentiment about too many floating patches. I suggest the following two process changes:
* Review changes before landing them on the `vee-eight-*` branches.
* Require that V8 back-ports be submitted first as merge-requests to upstream. Not just on the `vee-eight-*` branches, but also on `master`. Doing this as soon as possible is best.

@bnoordhuis I have dropped the `deps/v8/src/debug/mirrors.js` commit from 12581e6 as per your suggestion (please re-review). I have added back and fixed `test-debug-args.js` in cdb3936 (please re-review).

> @trevnorris: Also, is it necessary to have the deps: upgrade V8 to 4.6.85.23 commit?

The `4.6.85.23` commit from @targos brings in V8 4.6. My `4.6.85.25` commit is the application of the patches to bring us from `.23` to `.25`. Apart from aesthetics, is there another reason to avoid having one of these commits? It is going to be slightly more work for me to replace the `.23` commit and I am lazy :smile:. But if there is a good reason to do this, I can do it.

rvagg commented:
+1 to @domenic's concerns with one major diversion: there are clearly situations where it's much more in Node's interest to expedite changes than it is for V8. The two that I'm seeing here are:

1. Anything to do with Buffer performance, which is not as interesting for the V8 team but of acute interest to us, anything we can do to edge closer to pre io.js v3 performance levels I'm +1 on doing.
2. Things relating to postmortem which is more interesting in Node than it is in V8, particularly given our tooling ecosystem (well, mainly just mdb for now). We're playing catch-up with 0.10 and 0.12 here, so I'm +1 on the tiny changes @misterdjules has been backporting for this.

Beyond that, if the V8 team don't see it as urgent to back-port then I'd question us going out of the gate with it. Unfortunately we're going to be living with 4.6 for 6+ months so there's plenty of time to do backporting to fix stuff if it emerges over time that it really is important.

ofrobots commented:
We should not start with a presumption that V8 team is going to reject a merge-request for a Buffer or Post-mortem change (case in point: [post-mortem change back-ported by the V8 team](https://github.com/v8/v8/releases/tag/4.6.85.13)). We should at least request a merge back from upstream. Their disposition either way is helpful for us in deciding how to deal with the fix.

> Unfortunately we're going to be living with 4.6 for 6+ months

This is news to me. I thought semver-major changes can land on master, so I was expecting that 4.7 will land on master in ~6 weeks' time. Note that 4.7 isn't expected to have any [actionable change to the API](http://bit.ly/1ZBAG23).

rvagg commented:
> This is news to me. I thought semver-major changes can land on master, so I was expecting that 4.7 will land on master in ~6 weeks' time.

That's correct but `v5.x` will live for 6 months with this V8. We have yet to determine how `master` code will make its way to eager users' hands in the form of "canary" or other builds.

bnoordhuis commented:
LGTM

ofrobots commented:
`test-fs-symlink` [failed](https://ci.nodejs.org/job/node-test-commit-windows/937/nodes=win10/tapTestReport/test.tap-268/) on windows 10. I am not sure if it is flaky. Launched another CI: https://ci.nodejs.org/job/node-test-pull-request/509/

bnoordhuis commented:
Yes, it's flaky.  I've seen it fail a few times now.

ofrobots commented:
Thanks. Landed in af58aea.

Ping @misterdjules, in case you want to test post-mortem with V8 4.6.

misterdjules commented:
@ofrobots I had tested the vee-eight-4.6 branch regularly as changes were added. Tested again just now, it's all good, thank you for the heads up!

rvagg commented:
@misterdjules is that sign-off for v5 with this version of V8 with postmortem? I was going to make it a checklist item for v5.0.0 but it sounds like you're giving it the :thumbsup:

misterdjules commented:
@rvagg Interesting, I'm pretty confident that I had updated the [check list](https://github.com/nodejs/node/issues/3000#issue-107669195) to check "Prepare postmortem metadata in new V8 prior to 5.0.0" as indicated in [one of my comments](https://github.com/nodejs/node/issues/3000#issuecomment-146248949). I checked it again :)
